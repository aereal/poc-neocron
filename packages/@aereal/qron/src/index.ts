export * from "./dynamodb-sfn-task";
export * from "./transactional-task";
