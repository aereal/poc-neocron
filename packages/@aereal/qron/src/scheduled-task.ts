import { Construct } from "@aws-cdk/core";
import { ITable, AttributeType } from "@aws-cdk/aws-dynamodb";
import {
  StateMachine,
  Task,
  Choice,
  Condition,
  IChainable,
  Fail,
  Succeed,
} from "@aws-cdk/aws-stepfunctions";
import { Rule } from "@aws-cdk/aws-events";
import { SfnStateMachine } from "@aws-cdk/aws-events-targets";
import { AttributeValue, UpdateItemTask } from "./dynamodb-sfn-task";

export interface ScheduledTaskProps {
  readonly lockTable: ITable;
  readonly invokeMain: Task;
  readonly taskName: string;
  readonly invocationRule: Rule;
}

const keyTaskName = "taskName";

export class ScheduledTask extends Construct {
  constructor(scope: Construct, id: string, props: ScheduledTaskProps) {
    super(scope, id);

    const { lockTable, invokeMain, taskName, invocationRule } = props;

    const lockKey: AttributeValue = {
      name: keyTaskName,
      type: AttributeType.STRING,
      value: taskName,
    };

    const getLockResultPath = "$.Lock";

    const amount: AttributeValue = {
      name: ":amount",
      type: AttributeType.NUMBER,
      value: "1",
    };

    const getLock = new Task(this, "GetLock", {
      task: new UpdateItemTask({
        table: lockTable,
        parameters: {
          key: lockKey,
          updateExpression:
            "SET handledCount = if_not_exists(handledCount, :default) + :amount",
          expressionAttributeValues: [
            {
              name: ":default",
              type: AttributeType.NUMBER,
              value: "0",
            },
            amount,
          ],
          returnValues: "ALL_NEW",
        },
      }),
      resultPath: getLockResultPath,
    });

    const freeLock = (id: string): Task =>
      new Task(this, id, {
        task: new UpdateItemTask({
          table: lockTable,
          parameters: {
            key: lockKey,
            updateExpression: "SET handledCount = handledCount - :amount",
            expressionAttributeValues: [amount],
            returnValues: "ALL_NEW",
          },
        }),
      });

    // TODO: otherwise
    // TODO: retry?
    const checkLock = (next: IChainable): Choice =>
      new Choice(this, "CheckLock")
        .when(
          Condition.stringEquals(
            `${getLockResultPath}.Attributes.handledCount.N`,
            "1"
          ),
          next
        )
        .otherwise(freeLock("FailedLockFreed").next(new Fail(this, "Finite")));

    const stateMachine = new StateMachine(this, "StateMachine", {
      definition: getLock.next(
        checkLock(
          invokeMain.next(
            freeLock("SuccessFreeLock").next(new Succeed(this, "Succeed"))
          )
        )
      ),
    });

    invocationRule.addTarget(new SfnStateMachine(stateMachine));
  }
}
