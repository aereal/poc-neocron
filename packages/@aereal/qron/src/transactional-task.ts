import { Construct } from "@aws-cdk/core";
import { ITable, AttributeType } from "@aws-cdk/aws-dynamodb";
import {
  StateMachine,
  Task,
  Choice,
  Condition,
  IChainable,
  Fail,
  Succeed,
  IStateMachine,
} from "@aws-cdk/aws-stepfunctions";
import { AttributeValue, UpdateItemTask } from "./dynamodb-sfn-task";

export interface ScheduledTaskProps {
  readonly lockTable: ITable;
  readonly invokeMain: Task;
  readonly taskName: string;
}

const keyTaskName = "taskName";

export class TransactionalTask extends Construct {
  public readonly stateMachine: IStateMachine;

  constructor(scope: Construct, id: string, props: ScheduledTaskProps) {
    super(scope, id);

    const { lockTable, invokeMain, taskName } = props;

    const lockKey: AttributeValue = {
      name: keyTaskName,
      type: AttributeType.STRING,
      value: taskName,
    };

    const getLockResultPath = "$.Lock";

    const amount: AttributeValue = {
      name: ":amount",
      type: AttributeType.NUMBER,
      value: "1",
    };

    const getLock = new Task(this, "GetLock", {
      task: new UpdateItemTask({
        table: lockTable,
        parameters: {
          key: lockKey,
          updateExpression:
            "SET handledCount = if_not_exists(handledCount, :default) + :amount",
          expressionAttributeValues: [
            {
              name: ":default",
              type: AttributeType.NUMBER,
              value: "0",
            },
            amount,
          ],
          returnValues: "ALL_NEW",
        },
      }),
      resultPath: getLockResultPath,
    });

    const freeLock = (id: string): Task =>
      new Task(this, id, {
        task: new UpdateItemTask({
          table: lockTable,
          parameters: {
            key: lockKey,
            updateExpression: "SET handledCount = handledCount - :amount",
            expressionAttributeValues: [amount],
            returnValues: "ALL_NEW",
          },
        }),
      });

    // TODO: otherwise
    // TODO: retry?
    const checkLock = (next: IChainable): Choice =>
      new Choice(this, "CheckLock")
        .when(
          Condition.stringEquals(
            `${getLockResultPath}.Attributes.handledCount.N`,
            "1"
          ),
          next
        )
        .otherwise(freeLock("FailedLockFreed").next(new Fail(this, "Finite")));

    this.stateMachine = new StateMachine(this, "StateMachine", {
      definition: getLock.next(
        checkLock(
          invokeMain
            .addCatch(freeLock("AssumeLockFreed"))
            .next(
              freeLock("SuccessFreeLock").next(new Succeed(this, "Succeed"))
            )
        )
      ),
    });
  }
}
