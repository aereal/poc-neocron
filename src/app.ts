import { App, AppProps, Environment } from "@aws-cdk/core";

interface NeocronAppProps extends AppProps {
  readonly env: Environment;
}

export class NeocronApp extends App {
  static newFromContext = (): NeocronApp => {
    const { CDK_DEFAULT_ACCOUNT, CDK_DEFAULT_REGION } = process.env;
    if (CDK_DEFAULT_ACCOUNT === undefined) {
      throw new Error("default account not found");
    }
    if (CDK_DEFAULT_REGION === undefined) {
      throw new Error("default region not found");
    }
    return new NeocronApp({
      env: { account: CDK_DEFAULT_ACCOUNT, region: CDK_DEFAULT_REGION },
    });
  };

  private constructor(props: NeocronAppProps) {
    super(props);
  }
}
