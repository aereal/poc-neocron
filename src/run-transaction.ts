import { ExpressionBuilder } from "@aereal/cdk-dynamodb-expression";
import { ITable } from "@aws-cdk/aws-dynamodb";
import {
  Choice,
  Condition,
  Fail,
  INextable,
  State,
  StateMachineFragment,
  TaskStateBase,
} from "@aws-cdk/aws-stepfunctions";
import {
  DynamoAttributeValue,
  DynamoReturnValues,
  DynamoUpdateItem,
} from "@aws-cdk/aws-stepfunctions-tasks";
import { Construct } from "@aws-cdk/core";

export interface RunTransactionalTaskProps {
  /**
   * DynamoDB table manages tasks concurrency.
   * It must have partition key that's name is `taskName` and type is string.
   */
  readonly lockTable: ITable;

  /**
   * Main task state.
   */
  readonly invokeMain: TaskStateBase;

  /**
   * taskName is used as lock key so it must be unique in same [[lockTable]].
   */
  readonly taskName: string;
}

const keyTaskName = "taskName";

/**
 * RunTransaction executes given main task and manages its concurrency.
 *
 * Most common AWS services guarantee at-least-once execution but not exactly-once.
 *
 * The RunTransaction uses Step Function as task runner and DynamoDB as exclusive lock.
 *
 * So we can run some task that supported by Step Functions task in exactly-once manner if max concurrency is 1.
 */
export class RunTransactionalTask extends StateMachineFragment {
  public readonly startState: State;
  public readonly endStates: INextable[];
  private readonly getLockResultPath: string = "$.Lock";
  private readonly lockTable: ITable;
  private readonly taskName: string;
  private readonly amount: DynamoAttributeValue = DynamoAttributeValue.fromNumber(
    1
  );
  private readonly getLockExprBuilder: ExpressionBuilder;
  private readonly freeLockExprBuilder: ExpressionBuilder;

  constructor(scope: Construct, id: string, props: RunTransactionalTaskProps) {
    super(scope, id);

    const { lockTable, invokeMain, taskName } = props;
    this.lockTable = lockTable;
    this.taskName = taskName;
    this.getLockExprBuilder = new ExpressionBuilder();
    this.freeLockExprBuilder = new ExpressionBuilder();

    const getLockResultPath = "$.Lock";
    const assumeLockFreed = this.freeLockTask("AssumeLockFreed");
    const onSuccess = this.freeLockTask("SuccessFreeLock");
    const failedLockFreed = this.freeLockTask("FailedLockFreed");
    const invokeWithRecover = invokeMain
      .addCatch(assumeLockFreed)
      .next(onSuccess);
    const checkLock = new Choice(this, "CheckLock")
      .when(
        Condition.stringEquals(
          `${getLockResultPath}.Attributes.handledCount.N`,
          "1"
        ),
        invokeWithRecover
      )
      .otherwise(failedLockFreed.next(new Fail(this, "Finite")))
      .afterwards();
    const getLock = this.getLockTask();
    const definition = getLock.next(checkLock);
    this.startState = definition.startState;
    this.endStates = [...onSuccess.endStates, ...assumeLockFreed.endStates];
  }

  /**
   *
   */
  private getLockTask(): DynamoUpdateItem {
    const defaultCount = DynamoAttributeValue.fromNumber(0);
    const {
      expression,
      expressionAttributeNames,
      expressionAttributeValues,
    } = this.getLockExprBuilder
      .expr`SET handledCount = if_not_exists(handledCount, ${defaultCount}) + ${this.amount}`;
    return new DynamoUpdateItem(this, "GetLock", {
      table: this.lockTable,
      key: this.lockKey,
      updateExpression: expression,
      expressionAttributeNames,
      expressionAttributeValues,
      returnValues: DynamoReturnValues.ALL_NEW,
      resultPath: this.getLockResultPath,
    });
  }

  /**
   *
   */
  private freeLockTask(id: string): DynamoUpdateItem {
    const {
      expression,
      expressionAttributeValues,
      expressionAttributeNames,
    } = this.freeLockExprBuilder
      .expr`SET handledCount = handledCount - ${DynamoAttributeValue.fromNumber(
      1
    )}`;
    return new DynamoUpdateItem(this, id, {
      table: this.lockTable,
      key: this.lockKey,
      updateExpression: expression,
      expressionAttributeNames,
      expressionAttributeValues,
      returnValues: DynamoReturnValues.ALL_NEW,
    });
  }

  /**
   *
   */
  private get lockKey(): { [key: string]: DynamoAttributeValue } {
    return { [keyTaskName]: DynamoAttributeValue.fromString(this.taskName) };
  }
}
