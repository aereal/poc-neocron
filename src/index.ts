import { QronApp } from "./app";

QronApp.newFromContext().synth();
