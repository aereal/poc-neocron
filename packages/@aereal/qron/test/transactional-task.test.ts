import { Stack } from "@aws-cdk/core";
import { Table, AttributeType } from "@aws-cdk/aws-dynamodb";
import { Rule, Schedule } from "@aws-cdk/aws-events";
import { Task, TaskInput } from "@aws-cdk/aws-stepfunctions";
import { PublishToTopic } from "@aws-cdk/aws-stepfunctions-tasks";
import { Topic } from "@aws-cdk/aws-sns";
import { SynthUtils } from "@aws-cdk/assert";
import { TransactionalTask } from "../src";

describe("TransactionalTask", () => {
  test("ok", () => {
    const stack = new Stack();
    const lockTable = new Table(stack, "LockTable", {
      partitionKey: {
        name: "id",
        type: AttributeType.STRING,
      },
    });
    const topic = new Topic(stack, "Topic");
    new TransactionalTask(stack, "Task", {
      lockTable,
      taskName: "test-task",
      invocationRule: new Rule(stack, "RunEveryHourRule", {
        schedule: Schedule.cron({ minute: "0/10", weekDay: "MON-FRI" }),
      }),
      invokeMain: new Task(stack, "MainStack", {
        task: new PublishToTopic(topic, {
          message: TaskInput.fromObject({}),
        }),
      }),
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});
