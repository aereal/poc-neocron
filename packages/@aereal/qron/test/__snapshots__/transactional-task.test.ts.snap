// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionalTask add invocation rule 1`] = `
Object {
  "Resources": Object {
    "LockTableB9DACF42": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "RunEveryHourRule04806ED3": Object {
      "Properties": Object {
        "ScheduleExpression": "cron(0/10 * ? * MON-FRI *)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Ref": "TaskStateMachine55E3E11E",
            },
            "Id": "Target0",
            "RoleArn": Object {
              "Fn::GetAtt": Array [
                "TaskStateMachineEventsRole59CB3EFB",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "TaskStateMachine55E3E11E": Object {
      "DependsOn": Array [
        "TaskStateMachineRoleDefaultPolicy03E77486",
        "TaskStateMachineRole6225F761",
      ],
      "Properties": Object {
        "DefinitionString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"StartAt\\":\\"GetLock\\",\\"States\\":{\\"GetLock\\":{\\"Next\\":\\"CheckLock\\",\\"Parameters\\":{\\"TableName\\":\\"",
              Object {
                "Ref": "LockTableB9DACF42",
              },
              "\\",\\"Key\\":{\\"taskName\\":{\\"S\\":\\"test-task\\"}},\\"ExpressionAttributeValues\\":{\\":default\\":{\\"N\\":\\"0\\"},\\":amount\\":{\\"N\\":\\"1\\"}},\\"UpdateExpression\\":\\"SET handledCount = if_not_exists(handledCount, :default) + :amount\\",\\"ReturnValues\\":\\"ALL_NEW\\"},\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:aws:states:::dynamodb:updateItem\\",\\"ResultPath\\":\\"$.Lock\\"},\\"CheckLock\\":{\\"Type\\":\\"Choice\\",\\"Choices\\":[{\\"Variable\\":\\"$.Lock.Attributes.handledCount.N\\",\\"StringEquals\\":\\"1\\",\\"Next\\":\\"PublishTopic\\"}],\\"Default\\":\\"FailedLockFreed\\"},\\"FailedLockFreed\\":{\\"Next\\":\\"Finite\\",\\"Parameters\\":{\\"TableName\\":\\"",
              Object {
                "Ref": "LockTableB9DACF42",
              },
              "\\",\\"Key\\":{\\"taskName\\":{\\"S\\":\\"test-task\\"}},\\"ExpressionAttributeValues\\":{\\":amount\\":{\\"N\\":\\"1\\"}},\\"UpdateExpression\\":\\"SET handledCount = handledCount - :amount\\",\\"ReturnValues\\":\\"ALL_NEW\\"},\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:aws:states:::dynamodb:updateItem\\"},\\"Finite\\":{\\"Type\\":\\"Fail\\"},\\"PublishTopic\\":{\\"Next\\":\\"SuccessFreeLock\\",\\"Catch\\":[{\\"ErrorEquals\\":[\\"States.ALL\\"],\\"Next\\":\\"AssumeLockFreed\\"}],\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::sns:publish\\",\\"Parameters\\":{\\"TopicArn\\":\\"",
              Object {
                "Ref": "TopicBFC7AF6E",
              },
              "\\",\\"Message\\":{}}},\\"SuccessFreeLock\\":{\\"Next\\":\\"Succeed\\",\\"Parameters\\":{\\"TableName\\":\\"",
              Object {
                "Ref": "LockTableB9DACF42",
              },
              "\\",\\"Key\\":{\\"taskName\\":{\\"S\\":\\"test-task\\"}},\\"ExpressionAttributeValues\\":{\\":amount\\":{\\"N\\":\\"1\\"}},\\"UpdateExpression\\":\\"SET handledCount = handledCount - :amount\\",\\"ReturnValues\\":\\"ALL_NEW\\"},\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:aws:states:::dynamodb:updateItem\\"},\\"Succeed\\":{\\"Type\\":\\"Succeed\\"},\\"AssumeLockFreed\\":{\\"End\\":true,\\"Parameters\\":{\\"TableName\\":\\"",
              Object {
                "Ref": "LockTableB9DACF42",
              },
              "\\",\\"Key\\":{\\"taskName\\":{\\"S\\":\\"test-task\\"}},\\"ExpressionAttributeValues\\":{\\":amount\\":{\\"N\\":\\"1\\"}},\\"UpdateExpression\\":\\"SET handledCount = handledCount - :amount\\",\\"ReturnValues\\":\\"ALL_NEW\\"},\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:aws:states:::dynamodb:updateItem\\"}}}",
            ],
          ],
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "TaskStateMachineRole6225F761",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "TaskStateMachineEventsRole59CB3EFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TaskStateMachineEventsRoleDefaultPolicyB8563A93": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "TaskStateMachine55E3E11E",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TaskStateMachineEventsRoleDefaultPolicyB8563A93",
        "Roles": Array [
          Object {
            "Ref": "TaskStateMachineEventsRole59CB3EFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TaskStateMachineRole6225F761": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "states.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TaskStateMachineRoleDefaultPolicy03E77486": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "dynamodb:UpdateItem",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "LockTableB9DACF42",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "TopicBFC7AF6E",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TaskStateMachineRoleDefaultPolicy03E77486",
        "Roles": Array [
          Object {
            "Ref": "TaskStateMachineRole6225F761",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TopicBFC7AF6E": Object {
      "Type": "AWS::SNS::Topic",
    },
  },
}
`;

exports[`TransactionalTask ok 1`] = `
Object {
  "Resources": Object {
    "LockTableB9DACF42": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "TaskStateMachine55E3E11E": Object {
      "DependsOn": Array [
        "TaskStateMachineRoleDefaultPolicy03E77486",
        "TaskStateMachineRole6225F761",
      ],
      "Properties": Object {
        "DefinitionString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"StartAt\\":\\"GetLock\\",\\"States\\":{\\"GetLock\\":{\\"Next\\":\\"CheckLock\\",\\"Parameters\\":{\\"TableName\\":\\"",
              Object {
                "Ref": "LockTableB9DACF42",
              },
              "\\",\\"Key\\":{\\"taskName\\":{\\"S\\":\\"test-task\\"}},\\"ExpressionAttributeValues\\":{\\":default\\":{\\"N\\":\\"0\\"},\\":amount\\":{\\"N\\":\\"1\\"}},\\"UpdateExpression\\":\\"SET handledCount = if_not_exists(handledCount, :default) + :amount\\",\\"ReturnValues\\":\\"ALL_NEW\\"},\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:aws:states:::dynamodb:updateItem\\",\\"ResultPath\\":\\"$.Lock\\"},\\"CheckLock\\":{\\"Type\\":\\"Choice\\",\\"Choices\\":[{\\"Variable\\":\\"$.Lock.Attributes.handledCount.N\\",\\"StringEquals\\":\\"1\\",\\"Next\\":\\"PublishTopic\\"}],\\"Default\\":\\"FailedLockFreed\\"},\\"FailedLockFreed\\":{\\"Next\\":\\"Finite\\",\\"Parameters\\":{\\"TableName\\":\\"",
              Object {
                "Ref": "LockTableB9DACF42",
              },
              "\\",\\"Key\\":{\\"taskName\\":{\\"S\\":\\"test-task\\"}},\\"ExpressionAttributeValues\\":{\\":amount\\":{\\"N\\":\\"1\\"}},\\"UpdateExpression\\":\\"SET handledCount = handledCount - :amount\\",\\"ReturnValues\\":\\"ALL_NEW\\"},\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:aws:states:::dynamodb:updateItem\\"},\\"Finite\\":{\\"Type\\":\\"Fail\\"},\\"PublishTopic\\":{\\"Next\\":\\"SuccessFreeLock\\",\\"Catch\\":[{\\"ErrorEquals\\":[\\"States.ALL\\"],\\"Next\\":\\"AssumeLockFreed\\"}],\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::sns:publish\\",\\"Parameters\\":{\\"TopicArn\\":\\"",
              Object {
                "Ref": "TopicBFC7AF6E",
              },
              "\\",\\"Message\\":{}}},\\"SuccessFreeLock\\":{\\"Next\\":\\"Succeed\\",\\"Parameters\\":{\\"TableName\\":\\"",
              Object {
                "Ref": "LockTableB9DACF42",
              },
              "\\",\\"Key\\":{\\"taskName\\":{\\"S\\":\\"test-task\\"}},\\"ExpressionAttributeValues\\":{\\":amount\\":{\\"N\\":\\"1\\"}},\\"UpdateExpression\\":\\"SET handledCount = handledCount - :amount\\",\\"ReturnValues\\":\\"ALL_NEW\\"},\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:aws:states:::dynamodb:updateItem\\"},\\"Succeed\\":{\\"Type\\":\\"Succeed\\"},\\"AssumeLockFreed\\":{\\"End\\":true,\\"Parameters\\":{\\"TableName\\":\\"",
              Object {
                "Ref": "LockTableB9DACF42",
              },
              "\\",\\"Key\\":{\\"taskName\\":{\\"S\\":\\"test-task\\"}},\\"ExpressionAttributeValues\\":{\\":amount\\":{\\"N\\":\\"1\\"}},\\"UpdateExpression\\":\\"SET handledCount = handledCount - :amount\\",\\"ReturnValues\\":\\"ALL_NEW\\"},\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:aws:states:::dynamodb:updateItem\\"}}}",
            ],
          ],
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "TaskStateMachineRole6225F761",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "TaskStateMachineRole6225F761": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "states.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TaskStateMachineRoleDefaultPolicy03E77486": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "dynamodb:UpdateItem",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "LockTableB9DACF42",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "TopicBFC7AF6E",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TaskStateMachineRoleDefaultPolicy03E77486",
        "Roles": Array [
          Object {
            "Ref": "TaskStateMachineRole6225F761",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TopicBFC7AF6E": Object {
      "Type": "AWS::SNS::Topic",
    },
  },
}
`;
